app/com/linkedin/drelephant/exceptions/core/ExceptionFingerprintingSpark.java

```diff
@@ -0,0 +1,368 @@
+    String nodeManagerPort = configuration.get(NODEMANAGER_ADDRESS).split(HOSTNAME_PATTERN)[1];
+    String amHostHTTPAddress = parseURL(analyticJob.getAmContainerLogsURL(), nodeManagerPort);
+    String completeURLToQuery = MessageFormat.format(JOBHISTORY_ADDRESS_FOR_LOG, jobHistoryAddress, amHostHTTPAddress);
+    logger.info(" Query this url for error details " + completeURLToQuery);
+    return completeURLToQuery;
+  }
+
+  private String parseURL(String url, String nodeManagerPort) throws Exception {
+    String amAddress = null, containerID = null, userName = null;
+    String data[] = url.split(URL_PATTERN);
+    for (int idx = STARTING_INDEX_FOR_URL; idx < data.length; idx++) {
+      if (data[idx].contains(HOSTNAME_PATTERN)) {
+        amAddress = data[idx].split(HOSTNAME_PATTERN)[0] + HOSTNAME_PATTERN + nodeManagerPort;
+      } else if (data[idx].toLowerCase().contains(CONTAINER_PATTERN)) {
+        containerID = data[idx];
+      } else if (idx == data.length - 1) {
+        userName = data[idx];
+      }
+    }
+    return amAddress + URL_PATTERN + containerID + URL_PATTERN + containerID + URL_PATTERN + userName;
+  }
+
+  private void driverLogProcessingForException(BufferedReader in, List<ExceptionInfo> exceptions) throws IOException {
+    String inputLine;
+    while ((inputLine = in.readLine()) != null) {
+      if (inputLine.length() <= THRESHOLD_LOG_LINE_LENGTH.getValue() && isExceptionContains(inputLine)) {
+        if (debugEnabled) {
+          logger.debug(" ExceptionFingerprinting " + inputLine + "\t" + inputLine);
+        }
+        String exceptionName = inputLine;
+        int stackTraceLine = NUMBER_OF_STACKTRACE_LINE.getValue();
+        StringBuffer stackTrace = new StringBuffer();
+        while (stackTraceLine >= 0 && inputLine != null) {
+          stackTrace.append(inputLine);
+          stackTraceLine--;
+          inputLine = in.readLine();
+        }
+        addExceptions((exceptionName + "" + stackTrace).hashCode(), exceptionName, stackTrace.toString(),
+            ExceptionInfo.ExceptionSource.DRIVER, exceptions);
+        if (inputLine == null) {
+          break;
+        }
+      }
+    }
+  }
+
+  @Override
+  public LogClass classifyException(List<ExceptionInfo> exceptionInformation) {
+    if (exceptionInformation != null && exceptionInformation.size() > 0) {
+      Classifier classifier = ClassifierFactory.getClassifier(ClassifierType.RULE_BASE_CLASSIFIER);
+      classifier.preProcessingData(exceptionInformation);
+      LogClass logClass = classifier.classify(exceptionInformation);
+      return logClass;
+    }
+    return null;
+  }
+
+  @Override
+  public boolean saveData(String jobExecId) throws Exception {
+    JobExecution jobExecution = JobExecution.find.where().eq(JobExecution.TABLE.jobExecId, jobExecId).findUnique();
+    if (jobExecution == null) {
+      logger.error(" Job Execution with following id doesn't exist " + jobExecId);
```

Comment:
Do we require to throw an exception here? It may happen in many cases where Spark jobs are not going through tuning framework. Should be an info log.

Reply:
Even if they are not going through , they should exist in job execution . If the above case happens , then its very peculiar (since job execution should be the superset) ,that's why it should thrown exception

Reply:
Currently we are getting Spark job only from the Azkaban. There are other schedulers like Appworx, EasyData, which are not integrated. Number of jobs might be less. Even from Azkaban people can choose not to call Dr. Elephant API. And in open source people might not be using auto tuning itself. I think we should either set it as a WARN or INFO log and not throw exception. Otherwise we would be getting a lot of false ERROR and exception logs.